src/Microsoft.OData.Client/Serialization/HttpClientRequestMessage.cs

```diff
@@ -64,19 +67,31 @@ public class HttpClientRequestMessage : DataServiceClientRequestMessage, ISendin
         /// <summary>True if SendingRequest2Event is currently invoked, otherwise false.</summary>
         private bool inSendingRequest2Event;
 
+        private bool _disposed = false;
+
         /// <summary>
         /// Constructor for HttpClientRequestMessage.
         /// Initializes the <see cref="HttpClientRequestMessage"/> class.
         /// The args.ActualMethod is the actual method. In post tunneling situations method will be POST instead of the specified verb method.
         /// The args.method is the specified verb method
         /// </summary>
-        /// </summary>
         public HttpClientRequestMessage(DataServiceClientRequestMessageArgs args) 
             : base(args.ActualMethod)
         {
             _messageStream = new MemoryStream();
-            _handler = new HttpClientHandler();
-            _client = new HttpClient(_handler, disposeHandler: true);
+
+            _clientHandlerProvider = args.HttpClientHandlerProvider;
+            if (_clientHandlerProvider == null)
+            {
+                _handler = new HttpClientHandler();
+                _client = new HttpClient(_handler, disposeHandler: true);
+            }
+            else
+            {
+                _handler = _clientHandlerProvider.GetHttpClientHandler();
+                _client = new HttpClient(_handler, disposeHandler: false);
+            }

```

Comment:
```suggestion
            var clientHandlerProvider = args.HttpClientHandlerProvider;
            if (clientHandlerProvider == null)
            {
                this._handler = new HttpClientHandler();
                this._client = new HttpClient(this._handler, disposeHandler: true);
            }
            else
            {
                try
                {
                    this._handler = clientHandlerProvider.GetHttpClientHandler();
                    this._client = new HttpClient(this._handler, disposeHandler: false);
                }
                catch
                {
                    this._messageStream.Dispose();
                    throw;
                }
            }
```

The `HttpClient` constructor will throw in case of a `null` handler, plus we likely don't own the provider code, so it could throw anything really.